Refactor `User_UploadAvatar` (front-end)

jasperhuangg opened the pull request:
### Details
Since we changed `UserAPI::uploadAvatar` to return an `avatar` and `avatarThumbnail` [here](https://github.com/Expensify/Web-Expensify/pull/34469/files#diff-18e774f476c45b2129f5fa568785057429d32d026fef2b90312f6895db2c636eR2566-R2567), we need to change how they are accessed in NewDot once the API call returns.

This also standardizes any accessing of avatars with `avatar` for the high resolution image and `avatarThumbnail` for the low resolution image. For personal details, we use both high and low resolution images. For policies, we only use the high resolution image. For more information, see my comment [here](https://github.com/Expensify/App/pull/10235#discussion_r938215957).

Finally, this also updates avatar picker on the profile page and workspace settings page to be offline first, and use the new API commands.

### Fixed Issues
$ https://github.com/Expensify/Expensify/issues/214156

### Tests
1. Update your `.env` file to use the production API by following [this SO](https://stackoverflow.com/c/expensify/questions/11993). This needs to be done since our s3 bucket uploads only work in prod.
2. Sign into any account and open your Profile Settings (click on your avatar > Profile).
3. Click on the avatar picker and upload an avatar. Verify that no loading indicator appears, and that your avatar updates immediately after you click "Save" (the "Save" button in the cropping modal). Hard refresh (`<Command><Shift><R>`) the page, and verify that the new avatar is persisted.
    <img width="450" alt="Screen Shot 2022-08-26 at 9 54 31 AM" src="https://user-images.githubusercontent.com/31285285/186799795-e5f210f9-8c3d-43d5-9f0c-34ab66d3bcac.png">

4. Be ready to turn off your device wifi very quickly. Repeat step 3, but IMMEDIATELY turn off your device wifi before the avatar can finish uploading. It should still show the new avatar, but it should be greyed out like so:
    <img width="370" alt="Screen Shot 2022-08-26 at 9 59 10 AM" src="https://user-images.githubusercontent.com/31285285/186800353-bbb60c8a-3eb6-4df8-b2d7-d36ce77b530d.png">

5. Turn your device wifi back on, and verify the upload completes and the avatar isn't greyed out anymore.
6. Now click on the avatar picker again and select "Remove photo". Verify that the avatar is replaced with a default avatar. Reload the page and verify that this is persisted.
7. To test the error displaying, run the following code-snippet in the JS console, and verify it displays correctly:
    ```
    Onyx.merge('personalDetails', {
        '<current user email>': {
            errorFields: {
                avatar: {
                    1: 'Your profile picture could not be uploaded. Please try again later when you are online.'
                }
            }
        }
    });
    ```
    <img width="378" alt="Screen Shot 2022-08-30 at 11 36 59 AM" src="https://user-images.githubusercontent.com/3885503/187403683-d62ab2c2-fdff-45dd-acb6-5455a218708b.png">

8. Now repeat steps 3-6, but with the avatar picker in the Workspace Settings page.

### PR Review Checklist
<!--
This is a checklist for PR authors & reviewers. Please make sure to complete all tasks and check them off once you do, or else Expensify has the right not to merge your PR!
-->
#### Contributor (PR Author) Checklist
- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] iOS / native
    - [x] Android / native
    - [x] iOS / Safari
    - [x] Android / Chrome
    - [x] MacOS / Chrome
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] Any functional components have the `displayName` property
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

<details>
<summary><h4>PR Reviewer Checklist</h4>

The Contributor+ will copy/paste it into a new comment and complete it after the author checklist is completed
</summary>

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] iOS / native
    - [ ] Android / native
    - [ ] iOS / Safari
    - [ ] Android / Chrome
    - [ ] MacOS / Chrome
    - [ ] MacOS / Desktop
- [ ] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] Any functional components have the `displayName` property
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

</details>

### QA Steps
Same as the testing steps, except you can skip step 1 and 7.

- [ ] Verify that no errors appear in the JS console

### Screenshots
<!-- Add screenshots for all platforms tested. Pull requests won't be merged unless the screenshots show the app was tested on all platforms.-->

#### Web
<!-- Insert screenshots of your changes on the web platform-->

#### Mobile Web
<!-- Insert screenshots of your changes on the web platform (from a mobile browser)-->

#### Desktop
<!-- Insert screenshots of your changes on the desktop platform-->

#### iOS
<!-- Insert screenshots of your changes on the iOS platform-->

#### Android
<!-- Insert screenshots of your changes on the Android platform-->

jasperhuangg committed:
change to access via avatarURL instead of s3url

Beamanator commented on src/libs/actions/PersonalDetails.js:
```diff
@@ -331,7 +331,7 @@ function setAvatar(file) {
             }
 
-            setPersonalDetails({avatar: response.s3url, avatarUploading: false});
+            setPersonalDetails({avatar: response.avatarURL, avatarUploading: false});
```
Do you thinkkk this is a place we could also update `avatarHighResolution`? I guess my question is really: do we **have to** update both the high & low res avatar URLs here, and if not, "should we" to make things more clear?

jasperhuangg replied:
I think we probably should also update avatarHighResolution, it looks like the OptionsList needs it to display the most up-to-date avatars. It also just makes sense to do that since we're going to have both URLs anyways.

https://github.com/Expensify/App/blob/3d015423f360108205be4c81876ba4f4c68f34c1/src/libs/OptionsListUtils.js#L113-L127

https://github.com/Expensify/App/blob/3d015423f360108205be4c81876ba4f4c68f34c1/src/libs/ReportUtils.js#L340-L351

jasperhuangg replied:
I'm starting to get a bit confused because it looks like we follow two different naming conventions for the high/los-res avatar urls:

Here is one of the places we use `avatar` and `avatarHighResolution`
https://github.com/Expensify/Web-Expensify/blob/4fb3c863ddadc87b1387d6047146c10a9d4f4fc5/lib/AppInit.php#L55-L55

Whereas here we use `avatar` and `avatarThumbnail`
https://github.com/Expensify/App/blob/3d015423f360108205be4c81876ba4f4c68f34c1/src/libs/actions/PersonalDetails.js#L102-L103

I'm thinking we should probably just standardize the naming everywhere (at least the parts that are used by App) if we're planning on updating this.

jasperhuangg replied:
I guess the question is, what are the names we'd like to assign to the low and high resolution avatar URLs?

I personally think `avatar` for low-res and `avatarHighResolution` make more sense, since in App we use `avatar` most frequently (to display the smaller avatars in reports), and we use `avatarHighResolution` only when someone clicks on report details. Going to change to use those two for now and we can change again later if you disagree.

@Beamanator

jasperhuangg replied:
To add to the confusion, it appears policies also have a different naming convention for their avatarURLs. I'm guessing policies were implemented before we also started storing thumbnails, so everywhere we reference avatars returned with policies or policy summaries they're called `avatarURL` (which is basically `avatarHighResolution`):

https://github.com/Expensify/App/blob/3d015423f360108205be4c81876ba4f4c68f34c1/src/libs/actions/Policy.js#L94-L95

https://github.com/Expensify/App/blob/3d015423f360108205be4c81876ba4f4c68f34c1/src/pages/workspace/WorkspaceSettingsPage.js#L40-L41

jasperhuangg replied:
So to summarize, there are three ways we interact with avatar urls, I think we should definitely standardize naming across all three since they all touch `UserAPI::uploadAvatar` (and quite frankly it's really confusing to follow)

For PersonalDetail objects, `avatarHighResolution` is stored as `avatar` (this will ultimately be removed later on and added in the Web-E PR):
https://github.com/Expensify/App/blob/f3c831dc85259143a841faa9fcd4e020c7dd65cc/src/libs/actions/PersonalDetails.js#L314-L336. 
Also, we randomly store `avatarThumbnail` in PersonalDetails objects whenever we call `json_encode($personalDetailsObject)` in PHP:
https://github.com/Expensify/Web-Expensify/blob/4fb3c863ddadc87b1387d6047146c10a9d4f4fc5/lib/PersonalDetail.php#L184-L191

For Policies we store `avatarHighResolution` as `avatarURL`.
https://github.com/Expensify/App/blob/f3c831dc85259143a841faa9fcd4e020c7dd65cc/src/libs/actions/Policy.js#L409-L430
3.

jasperhuangg replied:
@Beamanator Actually going to wait on doing this since it's a relatively big change. Gonna ask in #engineering-chat.

Beamanator replied:
Dude good call on all of this needing to be updated & cleaned up!

Beamanator commented on src/libs/actions/Policy.js:
```diff
@@ -418,8 +418,8 @@ function uploadAvatar(policyID, file) {
                 // Update the policy with the new avatarURL as soon as we get it
-                Onyx.merge(`${ONYXKEYS.COLLECTION.POLICY}${policyID}`, {avatarURL: response.s3url, isAvatarUploading: false});
-                update(policyID, {avatarURL: response.s3url}, true);
+                Onyx.merge(`${ONYXKEYS.COLLECTION.POLICY}${policyID}`, {avatarURL: response.avatarURL, isAvatarUploading: false});
+                update(policyID, {avatarURL: response.avatarURL}, true);
```
Same questions here, about high res vs not high res avatars

jasperhuangg committed:
use avatar/avatarThumbnail for high/los res avatars

jasperhuangg committed:
use avatar/avatarThumbnail for high/low res avatars, add fallbacks for old deprecated avatar access

jasperhuangg committed:
remove unused

jasperhuangg committed:
style

jasperhuangg committed:
Merge branch 'main' of github.com:Expensify/App into jasper-uploadAvatarURLNameChange

jasperhuangg committed:
merge main and the App/10315

jasperhuangg committed:
add defaults for deploy

jasperhuangg commented on src/libs/actions/PersonalDetails.js:
```diff
@@ -329,7 +329,12 @@ function setAvatar(file) {
-            setPersonalDetails({avatar: response.s3url, avatarUploading: false});
+            let {avatar, avatarThumbnail} = response;
+
+            // TODO: remove the following 2 lines once https://github.com/Expensify/Web-Expensify/pull/34469 is merged, also change above access to const
+            avatar = avatar || response.s3url;
+            avatarThumbnail = avatarThumbnail || response.s3url;
+            setPersonalDetails({avatar, avatarThumbnail, avatarUploading: false}, false);
```
Since `User_UploadAvatar` currently still just returns `s3url` in the response, I've added a default here so that we don't have to worry about deploying this and https://github.com/Expensify/Web-Expensify/pull/34469 (which changes it to return `avatar` and `avatarThumbnail`) simultaneously.

jasperhuangg commented on src/libs/actions/Policy.js:
```diff
@@ -417,9 +420,13 @@ function uploadAvatar(policyID, file) {
+                // Update the policy with the new avatar as soon as we get it
+                let {avatar} = response;
+
+                // TODO: remove the following line once https://github.com/Expensify/Web-Expensify/pull/34469 is merged, also change above access to const
+                avatar = avatar || response.s3url;
+                Onyx.merge(`${ONYXKEYS.COLLECTION.POLICY}${policyID}`, {avatar, isAvatarUploading: false});
+                update(policyID, {avatar}, true);
```
Same as https://github.com/Expensify/App/pull/10235/files#r942703914

jasperhuangg committed:
fix usages of avatarURL

Beamanator commented on src/libs/actions/PersonalDetails.js:
```diff
@@ -329,7 +329,12 @@ function setAvatar(file) {
-            setPersonalDetails({avatar: response.s3url, avatarUploading: false});
+            let {avatar, avatarThumbnail} = response;
+
+            // TODO: remove the following 2 lines once https://github.com/Expensify/Web-Expensify/pull/34469 is merged, also change above access to const
```
I guess this link is now https://github.com/Expensify/Web-Expensify/pull/34346?

Beamanator commented on src/libs/actions/Policy.js:
```diff
@@ -90,8 +90,11 @@ function getSimplifiedPolicyObject(fullPolicyOrPolicySummary, isFromFullPolicy)
+        // avatar will be nested within the key "value"
+        avatar: fullPolicyOrPolicySummary.avatar
+            || lodashGet(fullPolicyOrPolicySummary, 'value.avatar', '')
+            || fullPolicyOrPolicySummary.avatarURL
+            || lodashGet(fullPolicyOrPolicySummary, 'value.avatarURL', ''),
```
Just making sure, if we still have to check `avatarUrl` or `value.avatarURL` should we keep that in the function doc? Looks like you removed it from the function docs

Beamanator commented on src/libs/actions/Policy.js:
```diff
@@ -417,9 +420,13 @@ function uploadAvatar(policyID, file) {
+                // Update the policy with the new avatar as soon as we get it
+                let {avatar} = response;
+
+                // TODO: remove the following line once https://github.com/Expensify/Web-Expensify/pull/34469 is merged, also change above access to const
```
Same, now https://github.com/Expensify/Web-Expensify/pull/34346

Beamanator reviewed (commented):
Code changes look good, a few tiny comments - haven't tested yet (will test after you do & you add a few screenshots plz :D)

jasperhuangg commented:
@Beamanator Will do! 

I think I should be able to test this by just trying to setting avatars locally for workspaces and personal details using the main branch, and then checking out this branch and seeing if we're still able to load them and update them.

I'm currently having some issues with my VPN and ngrok which is preventing me from running the App anywhere, but I'll get on testing as soon as I figure this out 🥲

Beamanator commented:
Sounds good man! Oof good luck with VPN and ngrok :'( I don't thinkkk you need the VPN for App dev though, do you? I don't think I've needed it, I rarely rarely turn the VPN on (since it's annoying from where I'm located)

jasperhuangg committed:
Merge branch 'main' of github.com:Expensify/App into jasper-uploadAvatarURLNameChange

jasperhuangg commented:
@Beamanator based on my comment [here](https://github.com/Expensify/Expensify/issues/214156#issuecomment-1217024480) let's get the Web PR deployed first, that way I'll be able to test the front-end changes with working s3 buckets.

Beamanator reviewed (commented):
Some more tiny comments

jasperhuangg commented:
@Beamanator converting back to draft since I'm planning on adding all the front-end changes to this PR now too.

jasperhuangg committed:
merge main

jasperhuangg committed:
remove unused during merge

Beamanator commented:
@jasperhuangg cool let me know when you want me to start reviewing / testing!

jasperhuangg committed:
resolve merge conflicts

jasperhuangg committed:
add back function doc

jasperhuangg committed:
fix review comments

JmillsExpensify commented:
When do we expect this PR to come out of draft now that the web PR is out on production?

jasperhuangg commented:
@JmillsExpensify I got a bit distracted by a few other API refactor issues that I assigned myself while I was waiting for the Web PR to hit production. I've already signed off for today but I'm planning on focusing on nothing but this PR tomorrow so I can get it reviewed + tested with help from @Beamanator ASAP

JmillsExpensify commented:
Sounds like a plan, thanks @jasperhuangg

jasperhuangg committed:
add DotIndicatorMessage component

jasperhuangg committed:
update grammatically incorrect copy

jasperhuangg committed:
add ErrorUtils.getLatestErrorMessage

jasperhuangg committed:
DotIndicatorMessage styles

jasperhuangg committed:
update copy

jasperhuangg committed:
add updateUserAvatar

jasperhuangg committed:
Update AvatarWithImagePicker with offline-first patterns

jasperhuangg committed:
update ProfilePage to use new AvatarPicker

jasperhuangg committed:
resolve merge conflicts

jasperhuangg committed:
fix stuff messed up during merge

jasperhuangg committed:
use avatarThumbnail

jasperhuangg committed:
remove unused

jasperhuangg commented on src/languages/en.js:
```diff
@@ -520,7 +520,7 @@ export default {
         passwordNotSet: 'We were unable to set your new password. We have sent you a new password link to try again.',
         setPasswordLinkInvalid: 'This set password link is invalid or has expired. A new one is waiting for you in your email inbox!',
-        validatingAccount: 'Verifying account',
+        validateAccount: 'Verify account',
```
This text is used in the button you press after you click the validation link sent to your email to set your password.

jasperhuangg committed:
use OfflineWithFeedback instead

jasperhuangg committed:
apply updates to workspace settings page

jasperhuangg committed:
update translation

jasperhuangg committed:
fix jest unit tests

jasperhuangg committed:
cleanup/remove unused stuff

jasperhuangg committed:
remove unnecessary network prop

jasperhuangg commented:
@Expensify/design @Beamanator 

With the new design, there's currently no user feedback when an avatar is uploaded successfully, which could potentially be kind of confusing. I don't want to drag this PR on any longer than it needs to be, should I create a follow-up issue to include some sort of user feedback?

Here's how it looks at the moment:

https://user-images.githubusercontent.com/31285285/186833783-975ea9f6-e352-4dba-a65f-67fff991a64a.mov

jasperhuangg committed:
more cleanup

jasperhuangg committed:
fix linter error

jasperhuangg committed:
remove unused styles

jasperhuangg commented on src/components/AvatarWithImagePicker.js:
```diff
@@ -252,41 +241,23 @@ class AvatarWithImagePicker extends React.Component {
-                                                </>
-                                            )
-                                    }
+                                    <Tooltip absolute text={this.props.translate('avatarWithImagePicker.editImage')}>
+                                        <View style={[styles.smallEditIcon, styles.smallAvatarEditIcon]}>
+                                            <Icon
+                                                src={Expensicons.Camera}
+                                                width={variables.iconSizeSmall}
+                                                height={variables.iconSizeSmall}
+                                                fill={themeColors.iconReversed}
+                                            />
+                                        </View>
+                                    </Tooltip>
+                                    <PopoverMenu
+                                        isVisible={this.state.isMenuVisible}
+                                        onClose={() => this.setState({isMenuVisible: false})}
+                                        onItemSelected={() => this.setState({isMenuVisible: false})}
+                                        menuItems={this.createMenuItems(openPicker)}
+                                        anchorPosition={this.props.anchorPosition}
+                                    />
```
@Beamanator I've removed the loading indicator here since the design doc mentions this:

> Both of these old flows (updating user & policy avatars) currently have loading states which will be removed since when a user updates an avatar (user or workspace) we can store that image data locally while we wait for the server to upload the file & respond with the avatar URL!

However, looking back, I think it might be good to keep the loading indicator since it'll potentially mitigate my concerns [here](https://github.com/Expensify/App/pull/10235#issuecomment-1228092978).

Here's how the loading indicator used to look. Do we want to keep it? This means I'll have to reintroduce a loading state into Onyx.
<img width="398" alt="Screen Shot 2022-08-26 at 2 36 22 PM" src="https://user-images.githubusercontent.com/31285285/186838111-03f13dd2-6680-4e43-84b5-6f251e3f5776.png">


cc @Expensify/design

Beamanator replied:
Interesting - personally I think it's ok not to have feedback because:
1. The image technically does update immediately, which is enough feedback for me that it succeeded ("optimistically")
2. If it fails we do show a nice error message (or if we go offline)

I'm definitely willing to discuss more but like you said @jasperhuangg maybe in a different issue or follow-up since the design doc does specifically call out having no feedback (pattern A)

jasperhuangg replied:
Cool cool! Just wanted to double-check. Let's wait and see what @Expensify/design has to say, but your logic makes sense to me and I'm more than happy to leave it like this.

shawnborton replied:
I agree with @Beamanator - the loading spinner kind of defeats the purpose of optimistic actions too right?

Beamanator commented on src/components/OfflineWithFeedback.js:
```diff
@@ -45,6 +45,9 @@ const propTypes = {
     /** Additional style object for the error row */
     errorRowStyles: stylePropTypes,
 
+    /** Whether we should allow the user to close the error message */
+    isCloseable: PropTypes.bool,
+
```
Did you add this just because we didn't have an X icon in the design? Now that I think about it I think we might want to keep the X, just so the user can hide it (and kill that RBR) if they don't care about fixing their profile picture that moment - what do you think?

jasperhuangg replied:
Yeah I totally got rid of it because there wasn't any X in the design, don't mind adding it back since it simplifies things!

Beamanator commented on src/libs/actions/PersonalDetails.js:
```diff
@@ -91,26 +91,26 @@ function formatPersonalDetails(personalDetailsList) {
         const payPalMeAddress = details.expensify_payPalMeAddress || '';
         const phoneNumber = details.phoneNumber || '';
-        const avatarHighResolution = details.avatar || details.avatarThumbnail;
+        const avatar = details.avatar || details.avatarThumbnail || ReportUtils.getDefaultAvatar(login);
+        const avatarThumbnail = getAvatarThumbnail(details, sanitizedLogin);
```
I like this a lot more than what we had before 👍 👍

Beamanator commented on src/pages/workspace/WorkspaceSettingsPage.js:
```diff
@@ -138,8 +123,8 @@ class WorkspaceSettingsPage extends React.Component {
                                     anchorPosition={{top: 172, right: 18}}
-                                    isUsingDefaultAvatar={!this.state.previewAvatarURL}
-                                    onImageSelected={this.uploadAvatar}
+                                    isUsingDefaultAvatar={!this.props.policy.avatar}
```
Probably should use `lodashGet` here:

```suggestion
                                    isUsingDefaultAvatar={!lodashGet(this.props.policy, 'avatar', null)}
```

Beamanator commented on src/pages/workspace/WorkspaceSettingsPage.js:
```diff
@@ -138,8 +123,8 @@ class WorkspaceSettingsPage extends React.Component {
-                                    isUsingDefaultAvatar={!this.state.previewAvatarURL}
-                                    onImageSelected={this.uploadAvatar}
+                                    isUsingDefaultAvatar={!this.props.policy.avatar}
+                                    onImageSelected={file => Policy.updateWorkspaceAvatar(this.props.policy.id, file)}
```
And here

```suggestion
                                    onImageSelected={file => Policy.updateWorkspaceAvatar(lodashGet(this.props.policy, 'id', ''), file)}
```

Beamanator commented on src/pages/workspace/WorkspaceSettingsPage.js:
```diff
@@ -138,8 +123,8 @@ class WorkspaceSettingsPage extends React.Component {
-                                    onImageSelected={this.uploadAvatar}
+                                    isUsingDefaultAvatar={!this.props.policy.avatar}
+                                    onImageSelected={file => Policy.updateWorkspaceAvatar(this.props.policy.id, file)}
                                     onImageRemoved={this.removeAvatar}
```
Why not also just directly call the Policy action here since we're doing this elsewhere?

```suggestion
                                    onImageRemoved={() => Policy.deleteWorkspaceAvatar(lodashGet(this.props.policy, 'id')}
```

Beamanator reviewed (changes requested):
Left a few initial comments - lots are related so there's like 3 requested changes that related to 10 different sections of code. Overall it's looking super great! Once we discuss these I'll start testing (hopefully today, maybe on the weekend)

jasperhuangg commented on src/pages/SetPasswordPage.js:
```diff
@@ -99,7 +99,7 @@ class SetPasswordPage extends Component {
 
     render() {
-        const buttonText = !this.props.account.validated ? this.props.translate('setPasswordPage.validatingAccount') : this.props.translate('setPasswordPage.setPassword');
+        const buttonText = !this.props.account.validated ? this.props.translate('setPasswordPage.validateAccount') : this.props.translate('setPasswordPage.setPassword');
```
It honestly has nothing to do with it, I just noticed this when i was testing and though it would be nice to update the copy since it didn't make grammatical sense.

jasperhuangg committed:
Update src/pages/workspace/WorkspaceSettingsPage.js

Co-authored-by: Alex Beaman <alexbeaman@expensify.com>

jasperhuangg committed:
remove isCloseable and store errors in errorFields.avatar

jasperhuangg committed:
Merge remote-tracking branch 'origin/jasper-uploadAvatarURLNameChange' into jasper-uploadAvatarURLNameChange

jasperhuangg commented on src/pages/workspace/WorkspaceSettingsPage.js:
```diff
@@ -62,22 +60,9 @@ class WorkspaceSettingsPage extends React.Component {
     }
 
     removeAvatar() {
-        this.setState({previewAvatarURL: ''});
         Policy.deleteWorkspaceAvatar(this.props.policy.id);
     }
```
huh coulda sworn I did just that.. this PR has definitely gotten a bit too chonky 😅

Beamanator replied:
First time I've heard "chonky" 😆

jasperhuangg committed:
directly use Policy action in onImageRemoved callback

jasperhuangg committed:
Update src/pages/workspace/WorkspaceSettingsPage.js

Co-authored-by: Alex Beaman <alexbeaman@expensify.com>

jasperhuangg committed:
Update src/pages/workspace/WorkspaceSettingsPage.js

Co-authored-by: Alex Beaman <alexbeaman@expensify.com>

jasperhuangg committed:
use lodashGet

jasperhuangg commented:
@Beamanator addressed your comments! I removed the `isCloseable` logic and access errors from `errorFields.avatar` (in lieu of your comment on the Web PR [here](https://github.com/Expensify/Web-Expensify/pull/34692#discussion_r955817881))

jasperhuangg committed:
merge main

jasperhuangg committed:
remove unused

Beamanator reviewed (commented):
When testing locally (not in prod database), I did this:
1. uploaded avatar
2. Saw `PDF Box can not upload user asset[]` error
3. Clicked "Save"
4. Saw this console error:

<img width="1357" alt="Screen Shot 2022-08-29 at 4 29 28 PM" src="https://user-images.githubusercontent.com/3885503/187225286-3bc0cf49-6f61-4b81-9e6f-62021d468f02.png">

I think this has something to do with the Save button being clickable when it shouldn't be? Do you experience this?

jasperhuangg committed:
use lodashGet

jasperhuangg commented on src/libs/actions/PersonalDetails.js:
```diff
@@ -361,15 +376,32 @@ function deleteAvatar() {
 }
 
+/**
+ * Clear error and pending fields for the current user's avatar
+ */
+function clearAvatarErrors() {
```
I see where you're coming from, but all the other actions that clear errors and pendingFields are just called `clearXXXXErrors`, so I'm electing to keep this the same.

jasperhuangg commented:
> When testing locally (not in prod database), I did this:
> 
> 1. uploaded avatar
> 2. Saw `PDF Box can not upload user asset[]` error
> 3. Clicked "Save"
> 4. Saw this console error:
> 
> <img alt="Screen Shot 2022-08-29 at 4 29 28 PM" width="1357" src="https://user-images.githubusercontent.com/3885503/187225286-3bc0cf49-6f61-4b81-9e6f-62021d468f02.png">
> 
> I think this has something to do with the Save button being clickable when it shouldn't be? Do you experience this?

@Beamanator hmmm I can't reproduce, the button never becomes clickable for me:
https://user-images.githubusercontent.com/31285285/187338509-f1550766-3393-4029-b713-3bca0d04d37e.mov

Here's the condition we're using to decide whether the button should be disabled. Looking around in [ProfilePage.js](https://github.com/Expensify/App/blob/ad841c0a9379053e216854e7c0762714f8c4db07/src/pages/settings/Profile/ProfilePage.js#L186-L190) I can't see how the button would become enabled by uploading the avatar.

```
const isButtonDisabled = (currentUserDetails.firstName === this.state.firstName.trim())
            && (currentUserDetails.lastName === this.state.lastName.trim())
            && (lodashGet(currentUserDetails, 'timezone.selected') === this.state.selectedTimezone)
            && (lodashGet(currentUserDetails, 'timezone.automatic') === this.state.isAutomaticTimezone)
            && (currentUserDetails.pronouns === this.state.pronouns.trim());
```

Beamanator commented:
Also FYI may want to update the screenshot of the error with the X? :D

<img width="378" alt="Screen Shot 2022-08-30 at 11 36 59 AM" src="https://user-images.githubusercontent.com/3885503/187403683-d62ab2c2-fdff-45dd-acb6-5455a218708b.png">

jasperhuangg committed:
clear pendingFields on success

jasperhuangg commented:
@Beamanator figured it out, it's because we're not clearing pendingFields in the successData, so the next time you go offline pendingField.avatar is still `'update'` when it should be `null`.

Beamanator reviewed (changes requested):
It looks like removing the workspace avatar isn't working - when I refresh the page (hard refresh)

jasperhuangg commented:
> It looks like removing the workspace avatar isn't working - when I refresh the page (hard refresh)

Ah my b, it's because we're clearing out `avatarURL` rather than `avatar` [here](https://github.com/Expensify/Web-Expensify/blob/4d541035f2800abad7d17ad9eadbcb5915016bd4/lib/PolicyAPI.php#L395-L396). I'll open up a PR.

Beamanator commented:
aaah good catch @jasperhuangg - I guess we need to get that fix CP'd before merging this right?

jasperhuangg commented:
> aaah good catch @jasperhuangg - I guess we need to get that fix CP'd before merging this right?

Yup! HOLDing this PR on https://github.com/Expensify/Web-Expensify/pull/34724 for now.

Beamanator commented:
@jasperhuangg merged! let's get it CP'd! :D

Beamanator reviewed (approved):
After a few rounds of testing everything's looking great!

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/2955919301) to staging by @Beamanator in version: 1.1.94-3 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅
